File: skypydb/embeddings/core/get_embeddings_function.py
"""Embedding provider factory helpers."""

from typing import Any, Callable, List


def _validate_remaining_config(
    provider: str,
    config: dict,
) -> None:
    """Validate that no unsupported config keys are provided."""

    if config:
        unsupported_keys = ", ".join(sorted(config.keys()))
        raise ValueError(
            f"Unsupported embedding config keys for provider '{provider}': {unsupported_keys}"
        )

def get_embedding_function(
    provider: str = "ollama",
    **config: Any
) -> Callable[[List[str]], List[List[float]]]:
    """
    Get an embedding function from supported providers.

    Args:
        provider: Embedding provider (ollama, openai, sentence-transformers)
        **config: Provider-specific configuration
    """

    provider = provider.lower().strip().replace("_", "-")

    # ollama provider
    if provider == "ollama":
        from skypydb.embeddings.ollama import OllamaEmbedding

        model = config.pop("model", "mxbai-embed-large")
        base_url = config.pop("base_url", "http://localhost:11434")
        dimension = config.pop("dimension", None)
        _validate_remaining_config(provider, config)
        return OllamaEmbedding(model=model, base_url=base_url, dimension=dimension)

    # openai provider
    if provider == "openai":
        from skypydb.embeddings.openai import OpenAIEmbedding

        api_key = config.pop("api_key", None)
        model = config.pop("model", "text-embedding-3-small")
        base_url = config.pop("base_url", None)
        organization = config.pop("organization", None)
        project = config.pop("project", None)
        timeout = config.pop("timeout", None)
        dimension = config.pop("dimension", None)
        _validate_remaining_config(provider, config)
        return OpenAIEmbedding(
            api_key=api_key,
            model=model,
            base_url=base_url,
            organization=organization,
            project=project,
            timeout=timeout,
            dimension=dimension
        )

    # sentence-transformers provider
    if provider in {"sentence-transformers", "sentence-transformer"}:
        from skypydb.embeddings.sentence_transformers import SentenceTransformerEmbedding

        model = config.pop("model", "all-MiniLM-L6-v2")
        device = config.pop("device", None)
        normalize_embeddings = config.pop("normalize_embeddings", False)
        dimension = config.pop("dimension", None)
        _validate_remaining_config(provider, config)
        return SentenceTransformerEmbedding(
            model=model,
            device=device,
            normalize_embeddings=normalize_embeddings,
            dimension=dimension
        )
    raise ValueError(
        f"Unsupported embedding provider '{provider}'. "
        "Supported providers: ollama, openai, sentence-transformers."
    )

File: skypydb-js/src/embeddings/core/get_embeddings_function.ts
import type { EmbeddingFunction } from "../../types";
import { OllamaEmbedding } from "../ollama";
import { OpenAIEmbedding } from "../openai";
import { SentenceTransformerEmbedding } from "../sentence_transformers";
export { get_embedding } from "./get_embedding";

function validate_remaining_config(
  provider: string,
  config: Record<string, unknown>,
): void {
  const keys = Object.keys(config);
  if (keys.length > 0) {
    throw new Error(
      `Unsupported embedding config keys for provider '${provider}': ${keys.sort().join(", ")}`,
    );
  }
}

export function get_embedding_function(
  provider = "ollama",
  config: Record<string, unknown> = {},
): EmbeddingFunction {
  const normalized = provider.toLowerCase().trim().replace(/_/g, "-");
  const mutable = { ...config };

  if (normalized === "ollama") {
    const model = (mutable.model as string | undefined) ?? "mxbai-embed-large";
    const base_url =
      (mutable.base_url as string | undefined) ?? "http://localhost:11434";
    const dimension = mutable.dimension as number | undefined;
    delete mutable.model;
    delete mutable.base_url;
    delete mutable.dimension;
    validate_remaining_config(normalized, mutable);
    const provider_instance = new OllamaEmbedding({
      model,
      base_url,
      dimension,
    });
    return provider_instance.embed.bind(provider_instance);
  }

  if (normalized === "openai") {
    const api_key = mutable.api_key as string | undefined;
    const model =
      (mutable.model as string | undefined) ?? "text-embedding-3-small";
    const base_url = mutable.base_url as string | undefined;
    const organization = mutable.organization as string | undefined;
    const project = mutable.project as string | undefined;
    const timeout = mutable.timeout as number | undefined;
    const dimension = mutable.dimension as number | undefined;
    delete mutable.api_key;
    delete mutable.model;
    delete mutable.base_url;
    delete mutable.organization;
    delete mutable.project;
    delete mutable.timeout;
    delete mutable.dimension;
    validate_remaining_config(normalized, mutable);
    const provider_instance = new OpenAIEmbedding({
      api_key,
      model,
      base_url,
      organization,
      project,
      timeout,
      dimension,
    });
    return provider_instance.embed.bind(provider_instance);
  }

  if (normalized === "sentence-transformers") {
    const model = (mutable.model as string | undefined) ?? "all-MiniLM-L6-v2";
    const device = mutable.device as string | undefined;
    const normalize_embeddings =
      (mutable.normalize_embeddings as boolean | undefined) ?? false;
    const dimension = mutable.dimension as number | undefined;
    delete mutable.model;
    delete mutable.device;
    delete mutable.normalize_embeddings;
    delete mutable.dimension;
    validate_remaining_config(normalized, mutable);
    const provider_instance = new SentenceTransformerEmbedding({
      model,
      device,
      normalize_embeddings,
      dimension,
    });
    return provider_instance.embed.bind(provider_instance);
  }

  throw new Error(
    `Unsupported embedding provider '${provider}'. Supported providers: ollama, openai, sentence-transformers.`,
  );
}
